tokio/src/time/clock.rs

```diff
@@ -124,7 +124,17 @@ cfg_test_util! {
         let until = clock.now() + duration;
         clock.advance(duration);
 
-        crate::time::sleep_until(until).await;
+        // Fixes tokio-rs/tokio#3837
+        //
+        // "sleep_until" is needed to ensure registered timers are fired after the
+        // advance call, however, it will also round up time to the nearest ms. To
+        // avoid this loss of precision, we sleep until the previous millisecond.
+        // If the advance is less than 1ms, then we just need to yield.
+        if duration >= Duration::from_millis(1) {
+            crate::time::sleep_until(until - Duration::from_millis(1)).await;
+        } else {
+            crate::task::yield_now().await;
+        }
```

Comment:
Can we have a test that does the following:

 * Creates a timer with a duration of 1 ms.
 * Advances time by 0.25 ms.
 * Advances time by 1.50 ms.

and one that does this:

 * Creates a timer with a duration of 1 ms.
 * Advances time by 0.25 ms.
 * Advances time by 0.25 ms.
 * Advances time by 0.25 ms.
 * Advances time by 0.25 ms.

Reply:
Thanks, this uncovered a different bug.

Reply:
I added the tests.